include/materials/material_utility.tcc

```diff
@@ -0,0 +1,221 @@
+//! Compute mean stress p (compression positive)
+inline const double mpm::material::p(
+    const Eigen::Matrix<double, 6, 1>& stress) {
+  return (-1. / 3. * (stress(0) + stress(1) + stress(2)));
+}
+
+//! Compute deviatoric stress
+inline const Eigen::Matrix<double, 6, 1> mpm::material::deviatoric_stress(
+    const Eigen::Matrix<double, 6, 1>& stress) {
+
+  // Compute mean p in tension positive
+  const double p = -1.0 * mpm::material::p(stress);
+
+  // Compute deviatoric by subtracting volumetric part
+  Eigen::Matrix<double, 6, 1> deviatoric_stress = stress;
+  for (unsigned i = 0; i < 3; ++i) deviatoric_stress(i) -= p;
+
+  return deviatoric_stress;
+}
+
+//! Compute J2 invariant
+inline const double mpm::material::j2(
+    const Eigen::Matrix<double, 6, 1>& stress) {
+
+  const double j2 = (std::pow((stress(0) - stress(1)), 2) +
+                     std::pow((stress(1) - stress(2)), 2) +
+                     std::pow((stress(0) - stress(2)), 2)) /
+                        6.0 +
+                    std::pow(stress(3), 2) + std::pow(stress(4), 2) +
+                    std::pow(stress(5), 2);
+
+  return j2;
+}
+
+//! Compute J3 invariant
+inline const double mpm::material::j3(
+    const Eigen::Matrix<double, 6, 1>& stress) {
+
+  // Compute deviatoric stress
+  Eigen::Matrix<double, 6, 1> deviatoric_stress =
+      mpm::material::deviatoric_stress(stress);
+
+  // Compute J3
+  const double j3 =
+      (deviatoric_stress(0) * deviatoric_stress(1) * deviatoric_stress(2)) -
+      (deviatoric_stress(2) * std::pow(deviatoric_stress(3), 2)) +
+      ((2 * deviatoric_stress(3) * deviatoric_stress(4) *
+        deviatoric_stress(5)) -
+       (deviatoric_stress(0) * std::pow(deviatoric_stress(4), 2)) -
+       (deviatoric_stress(1) * std::pow(deviatoric_stress(5), 2)));
+
+  return j3;
+}
+
+//! Compute deviatoric q
+inline const double mpm::material::q(
+    const Eigen::Matrix<double, 6, 1>& stress) {
+
+  // Compute J2 from
+  const double j2 = mpm::material::j2(stress);
+
+  return (std::sqrt(3 * j2));
+}
+
+//! Compute Lode angle
+inline const double mpm::material::lode_angle(
+    const Eigen::Matrix<double, 6, 1>& stress) {
+
+  // Compute j2 and j3
+  const double j2 = mpm::material::j2(stress);
+  const double j3 = mpm::material::j3(stress);
+
+  // Compute Lode angle value
+  double lode_angle_val = 0.0;
+  if (std::abs(j2) > std::numeric_limits<double>::epsilon()) {
+    lode_angle_val = (3. * std::sqrt(3.) / 2.) * (j3 / std::pow(j2, 1.5));
+  }
+  if (lode_angle_val > 1.0) lode_angle_val = 1.0;
+  if (lode_angle_val < -1.0) lode_angle_val = -1.0;
+
+  // Compute Lode angle (sin convention, between 0 and pi/3)
+  double lode_angle = (1. / 3.) * acos(lode_angle_val);
+
+  return lode_angle;
+}
+
+//! Compute derivative of p in terms of stress sigma
+inline const Eigen::Matrix<double, 6, 1> mpm::material::dp_dsigma(
+    const Eigen::Matrix<double, 6, 1>& stress) {
+
+  Eigen::Matrix<double, 6, 1> dp_dsigma;
```

Comment:
Using set zero will be 2 steps every time, the way you had it was how we want it, but instead use the `zero<6>()` to initialize them to zeros.

Reply:
This function is available in `data_types.h`, you may want to specialize for `6`

Reply:
I see! Give me a moment

Reply:
I updated the `data_types.h` but is this how I use it?
`Eigen::Matrix<double, 6, 1> dp_dsigma = mpm::zero<6>();`? (It does not work)